sleap_nn/data/edge_maps.py

```diff
@@ -1 +1,368 @@
+            each edge.
+        sigma: Standard deviation of the 2D Gaussian distribution sampled to generate
+            confidence maps.
+
+    Returns:
+        A set of confidence maps corresponding to the probability of each point on a
+        sampling grid being on each edge. These will be in a tensor of shape
+        (grid_height, grid_width, n_edges) of dtype torch.float32.
+    """
+    yy, xx = torch.meshgrid(yv, xv, indexing="ij")
+    sampling_grid = torch.stack((xx, yy), dim=-1)  # (height, width, 2)
+
+    distances = distance_to_edge(
+        sampling_grid, edge_source=edge_source, edge_destination=edge_destination
+    )
+    edge_maps = gaussian_pdf(distances, sigma=sigma)
+    return edge_maps
+
+
+def make_pafs(
+    xv: torch.Tensor,
+    yv: torch.Tensor,
+    edge_source: torch.Tensor,
+    edge_destination: torch.Tensor,
+    sigma: float,
+) -> torch.Tensor:
+    """Generate part affinity fields for a set of directed edges.
+
+    Args:
+        xv: Sampling grid vector for x-coordinates of shape (grid_width,) and dtype
+            torch.float32. This can be generated by
+            `sleap.nn.data.utils.make_grid_vectors`.
+        yv: Sampling grid vector for y-coordinates of shape (grid_height,) and dtype
+            torch.float32. This can be generated by
+            `sleap.nn.data.utils.make_grid_vectors`.
+        edge_source: Tensor of dtype torch.float32 of shape (n_edges, 2) where the last
+            axis corresponds to x- and y-coordinates of the source points of each edge.
+        edge_destination: Tensor of dtype torch.float32 of shape (n_edges, 2) where the
+            last axis corresponds to x- and y-coordinates of the destination points of
+            each edge.
+        sigma: Standard deviation of the 2D Gaussian distribution sampled to generate
+            the edge maps for masking the PAFs.
+
+    Returns:
+        A set of part affinity fields corresponding to the unit vector pointing along
+        the direction of each edge weighted by the probability of each point on a
+        sampling grid being on each edge. These will be in a tensor of shape
+        (grid_height, grid_width, n_edges, 2) of dtype torch.float32. The last axis
+        corresponds to the x- and y-coordinates of the unit vectors.
+    """
+    unit_vectors = edge_destination - edge_source
+    unit_vectors = unit_vectors / torch.norm(unit_vectors, dim=-1, keepdim=True)
+    edge_confidence_map = make_edge_maps(
+        xv=xv,
+        yv=yv,
+        edge_source=edge_source,
+        edge_destination=edge_destination,
+        sigma=sigma,
+    )
+    pafs = torch.unsqueeze(edge_confidence_map, dim=-1) * expand_to_rank(
+        unit_vectors, 4
+    )
+    return pafs
+
+
+def make_multi_pafs(
+    xv: torch.Tensor,
+    yv: torch.Tensor,
+    edge_sources: torch.Tensor,
+    edge_destinations: torch.Tensor,
+    sigma: float,
+) -> torch.Tensor:
+    """Make multiple instance PAFs with max reduction.
```

Comment:
Technically a reduction through addition not max (this is the case for confidence maps and I think this was carried over by accident)